flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/FullChangelogStoreWriteOperator.java

```diff
@@ -0,0 +1,224 @@
+                                        .add(Tuple2.of(t.f1, t.f2)));
+
+        snapshotIdentifierToCheck = null;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        currentWrittenBuckets = new HashSet<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        SinkRecord record = writeRecord(element);
+
+        BinaryRowData partition = record.partition();
+        int bucket = record.bucket();
+        // partition is a reused BinaryRowData
+        // we first check if the tuple exists to minimize copying
+        if (!currentWrittenBuckets.contains(Tuple2.of(partition, bucket))) {
+            currentWrittenBuckets.add(Tuple2.of(partition.copy(), bucket));
+        }
+    }
+
+    @Override
+    public void snapshotState(StateSnapshotContext context) throws Exception {
+        super.snapshotState(context);
+
+        List<Tuple3<Long, BinaryRowData, Integer>> writtenBucketList = new ArrayList<>();
+        for (Map.Entry<Long, Set<Tuple2<BinaryRowData, Integer>>> entry :
+                writtenBuckets.entrySet()) {
+            for (Tuple2<BinaryRowData, Integer> bucket : entry.getValue()) {
+                writtenBucketList.add(Tuple3.of(entry.getKey(), bucket.f0, bucket.f1));
+            }
+        }
+        writtenBucketState.update(writtenBucketList);
+    }
+
+    @Override
+    protected List<Committable> prepareCommit(boolean blocking, long checkpointId)
+            throws IOException {
+        if (snapshotIdentifierToCheck != null) {
+            Optional<Snapshot> snapshot = findSnapshot(snapshotIdentifierToCheck);
+            if (snapshot.map(s -> s.commitKind() == Snapshot.CommitKind.COMPACT).orElse(false)) {
+                writtenBuckets.headMap(snapshotIdentifierToCheck, true).clear();
+                snapshotIdentifierToCheck = null;
+            }
+        }
+
+        if (!currentWrittenBuckets.isEmpty()) {
+            writtenBuckets
+                    .computeIfAbsent(checkpointId, k -> new HashSet<>())
+                    .addAll(currentWrittenBuckets);
+            currentWrittenBuckets.clear();
+        }
+
+        if (snapshotIdentifierToCheck == null // wait for last forced full compaction to complete
+                && !writtenBuckets.isEmpty() // there should be something to compact
+                && checkpointId - writtenBuckets.navigableKeySet().first() + 1
+                        >= checkpointThreshold // checkpoints without full compaction exceeds
+        ) {
+            blocking = true;
+        }
+
+        if (blocking) {
+            snapshotIdentifierToCheck = checkpointId;
+            Set<Tuple2<BinaryRowData, Integer>> compactedBuckets = new HashSet<>();
+            for (Set<Tuple2<BinaryRowData, Integer>> buckets : writtenBuckets.values()) {
+                for (Tuple2<BinaryRowData, Integer> bucket : buckets) {
+                    if (compactedBuckets.contains(bucket)) {
```

Comment:
Why we need a `compactedBuckets`? Deduplicate? Why?

Reply:
For deduplication. It is possible that the same bucket appears across multiple checkpoints.